src/deluge/model/mod_controllable/mod_controllable_audio.cpp

```diff
@@ -1412,15 +1414,156 @@ bool ModControllableAudio::offerReceivedCCToLearnedParams(MIDIDevice* fromDevice
+						else if (midiKnobPos == (knobPos + 1)) {
+							newKnobPos = knobPos + 1;
+						}
+
+						else {
+
+							//if the first two conditions fail and pickup mode is enabled, then the Deluge Knob Position (and therefore the Parameter Value with it) remains unchanged
+							if (midiEngine.midiTakeover == MIDI_TAKEOVER_PICKUP) {//(runtimeFeatureSettings.get(RuntimeFeatureSettingType::MidiTakeoverMode) == RuntimeFeatureStateToggle::PU) { //Midi Pickup Mode On
+								newKnobPos = knobPos;
+							}
+							//if the first two conditions fail and value scaling mode is enabled, then the Deluge Knob Position is scaled upwards or downwards based on relative
+							//positions of Midi Controller Knob and Deluge Knob to min/max of knob range.
+							else { //Midi Value Scaling Mode On
+								//Set the max and min of the deluge midi knob position range
+								int knobMaxPos = 64;
+								int knobMinPos = -64;
+
+								//calculate amount of deluge "knob runway" is remaining from current knob position to max and min of knob position range
+								int delugeKnobMaxPosDelta = knobMaxPos - knobPos; //calculate how far delugeKnobPos is from Max - Positive Runway
+								int delugeKnobMinPosDelta = knobPos - knobMinPos; //calculate how far delugeKnobPos is from Min - Negative Runway
+
+								//calculate amount of midi "knob runway" is remaining from current knob position to max and min of knob position range
+								int midiKnobMaxPosDelta = knobMaxPos - midiKnobPos; //calculate how far midiKnobPos is from Max - Positive Runway
+								int midiKnobMinPosDelta = midiKnobPos - knobMinPos; //calculate how far midiKnobPos is from Min - Negative Runway
+
+								//calculate by how much the current midiKnobPos has changed from the previous midiKnobPos recorded
+								int midiKnobPosChange = midiKnobPos - knob->previousPosition;
+
+								//Set fixed point variable which will be used calculate the percentage in midi knob position
+								int midiKnobPosChangePercentage;
+
+								//if midi knob position change is greater than 0, then the midi knob position has increased (e.g. turned knob right)
+								if (midiKnobPosChange > 0) {
+									//fixed point math calculation of new deluge knob position when midi knob position has increased
+
+									midiKnobPosChangePercentage = (midiKnobPosChange * 1000000) / midiKnobMaxPosDelta;
+
+									newKnobPos = knobPos + ((delugeKnobMaxPosDelta * midiKnobPosChangePercentage) / 1000000);
+								}
+								//if midi knob position change is less than 0, then the midi knob position has decreased (e.g. turned knob left)
+								else if (midiKnobPosChange < 0) {
+									//fixed point math calculation of new deluge knob position when midi knob position has decreased
+
+									midiKnobPosChangePercentage = (midiKnobPosChange * 1000000) / midiKnobMinPosDelta;
+
+									newKnobPos = knobPos + ((delugeKnobMinPosDelta * midiKnobPosChangePercentage) / 1000000);
```

Comment:
Style: use <<20 and >>20 instead of the multiplication/division by 1000000, it makes it clearer how many bits are being used for the integer and fractional parts


The performance change would be negligible here but it's better practice

Reply:
I have made the recommended change and replaced the multiplication/division with << 20 and >> 20 respectively. See most recent commit.